### PR TITLE
(#692) Fix duration mismatch between audio and metadata

### LINKED ISSUE #692: Audio waveform progress bar extends beyond the end of the waveform body
## Description

When playing an audio file all the way through to the end, the progress bar will extend beyond the end of the waveform body.

I'm not sure what causes this, but it seeems like it would just be a small math issue or more likely we're ignoring something critical when pulling the [waveform dimensions from `clientWidth`](https://github.com/WordPress/openverse-frontend/blob/0ebfbe66c923c548954b38051d1dcad4fd54b539/src/components/VAudioTrack/VWaveform.vue#L294).

## Reproduction

1. Run main and go to an audio search that includes a short file (it's easiest and most prominent on short files) like http://localhost:8443/audio/5365bca5-7bd8-4344-b8e4-dad1cf219456
2. Play the audio file through to the end
3. Note the progress bar, timestamp, and focus bar (when focused) extending beyond the end of the waveform body.

## Expectation

The progress bar should end exactly at the end of the waveform body.

## Screenshots

- [ ] 🙋 I would be interested in resolving this bug.

### PR DESCRIPTION
**Pull request by demophoon**
When an audio file is played and the duration from the passed properties
does not match the duration of the actual audio file, e.g. the duration
was rounded, the progress bar drifts outside of the total width of the
VWaveform dimensions.

On smaller width screens or larger tracks it is less noticeable but when
the audio track is small and the screen width is large that percentage
adds up to 10s or more of pixels which is where the bug crops up.

This commit fixes this issue by by using the audio context's duration
parameter and falling back to less specific sources for duration.

## Fixes

Fixes #692 by @sarayourfriend 

## Checklist

- [x] My pull request has a descriptive title (not a vague title like `Update index.md`).
- [x] My pull request targets the *default* branch of the repository (`main`) or a parent feature branch.
- [ ] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [ ] I added or updated tests for the changes I made (if applicable).
- [ ] I added or updated documentation (if applicable).
- [ ] I tried running the project locally and verified that there are no visible errors.

[best_practices]:https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

File: src/components/VAudioTrack/VAudioTrack.vue [hunk 1]
```diff
@@ -285,7 +285,9 @@ export default defineComponent({
 
     /* Timekeeping */
 
-    const duration = computed(() => (props.audio?.duration ?? 0) / 1e3) // seconds
+    const duration = computed(
+      () => (localAudio?.duration ?? props.audio?.duration ?? 0) / 1e3 // seconds
```
Comment (sarayourfriend) [inline src/components/VAudioTrack/VAudioTrack.vue:289]: This looks great! There's also another component that will need this update: `VGlobalAudioTrack`.

You can see this component on the audio search results page when your viewport is small, like on a mobile device (or just a narrow desktop browser window).

Comment (demophoon) [reply src/components/VAudioTrack/VAudioTrack.vue:289]: Updated in `f3aa9f2`.

Hope its okay I added this to the single commit. Figured it was the same exact fix.

**Comment by demophoon** [comment]
Hmm.. this actually needs more investigation before being merged. I'm still seeing the bar fall off the end..

**Comment by sarayourfriend** [comment]
I think you have the right fit, but because the `localAudio` object isn't a `ref`, it's not going to recalculate.

One potential solution I can think of is to create a `duration` ref to store it in and use the `props.audio.duration` as the default value. Then we can add a `load` event listener to the `localAudio` object to retrieve the `duration` of the file once it's loaded.

What do you think @demophoon?

**Comment by zackkrida** [comment]
@demophoon do you think you'll have an opportunity to revisit this PR? If not, we can get another contributor to look at it, totally up to you. Thank you for your work so far!

**Comment by zackkrida** [comment]
@dhruvkb's expertise here would be useful as well, as the primary author of the original audio track components.

**Comment by demophoon** [comment]
@zackkrida Oh yes! It totally slipped out of my mind after the conference. I'm looking at this again this morning.

@sarayourfriend That makes total sense to me since the duration that I initially used is set before the browser has even had a chance to decode the audio track to determine the real duration.

**Comment by demophoon** [comment]
That change appears to have worked!

If there is a better way to notify a computed property of this change please let me know. I'm still a new to vue so I might not be aware of all the features.

Screenshots below of fix!
image
image

**Comment by sarayourfriend** [comment]
Looking good! Seems to work well on the single audio results, but for the global player there's a subtle bug.

First, to activate the global audio play, open a very narrow browser window on the audio search results page, something like this:

image

When you play an audio file, the global audio player will appear floating at the bottom of the screen. Make sure to narrow the results to just Jamendo provider (for example `/search/audio?q=birds&source=jamendo`).

Next, play one audio file that you know is shorter than another. Then play another audio file. Confirm that the duration was updated for the global player. Next, play the first audio again.

Notice that the duration will not be updated in the global player's context.

This is because the `durationchanged` event isn't firing, the audio objects have just been swapped out.

I think in the global audio track component we need to use the `durationchanged` event _as well as_ grabbing the duration when the `activeTrack` ref changes, similar to how we grab the `currentTime` of the active audio on this line:

https://github.com/WordPress/openverse-frontend/blob/fc35e75208b67ae022c41828997a6735f334af1f/src/components/VAudioTrack/VGlobalAudioTrack.vue#L157

In fact, I think just adding another line below that one that does this would work:

```js
audioDuration.value = audio.duration
```

I tried that locally and it seemed to work :+1:

**Comment by demophoon** [comment]
Oh, great catch. I didn't think about the object still being around on a second play through especially with respect to the global audio component.

Updated in `9f228640`.

**Review by sarayourfriend — APPROVED** [review]
This is great! Thanks so much for the contribution :tada:

**Review by zackkrida — APPROVED** [review]
So nice to see this fixed. The new code for `durationchange` listening and storage in a ref seems very logical!

**Pull request merged**